### PR TITLE
feat(core/onboarding): support manual rebase/retry

### LINKED ISSUE #11770: Add retry/rebase checkbox to the onboarding (configure renovate) pull requests
### What would you like Renovate to be able to do?

Add retry/rebase checkbox to the onboarding (configure renovate) pull requests

### If you have any ideas on how this should be implemented, please tell us here.

Add retry/rebase checkbox to the onboarding (configure renovate) pull requests

### Is this a feature you are interested in implementing yourself?

No

### LINKED ISSUE #17152: Support rebase/retry checkbox for onboarding PRs
### What would you like Renovate to be able to do?

Support rebase/retry checkboxes on onboarding PRs. This way we can reduce the frequency at which we check them for updates, because users will have the ability to request it be refreshed on-demand.

### If you have any ideas on how this should be implemented, please tell us here.

Same text as we have for regular PRs. But we also need to make sure our logic is in place to do the refresh/force push.

### Is this a feature you are interested in implementing yourself?

No

### PR DESCRIPTION
**Pull request by Gabriel-Ladzaretti**
## Changes
- Additional feature flag to enable onboarding rebase checkbox
- Added static onboarding state holder for manual retry request

Based on the feature flag value - 
- Onboarding PRs body will now contain embedded config hash + rebase checkbox 

Supported platforms -
- GitHub
- GitLab
- Gitea

Unsupported platforms will be unaffected by this change.

Changed logic -
`Extraction -> lookup -> PR creation/update` flow will now be preformed only when one of the following condition is met:
- No onboarding PR exists 
- Missing rebase checkbox
- Manual retry requested
- Branch is behind base
- Onboarding config was modified by the user (config hash changed)
- Missing PR embedded config hash

If non of the above is met, repo cache is left unchanged as well (`setBranchCache` is skipped as part of `worker/index`, `cache.scan` is written as part of the extraction, therefore is skipped as well).

Note that the rebase terminology in the context of onboarding is a bit misleading, as the actual action preformed by checking the checkbox is Updating the onboarding PR.

Design consideration -
- Embedding config hash as opposed to repo cache it: 
   - We currently sample the onboarding PR
   - Onboarding is the only case where the renovate config is present in a non base branch, caching it would pollute the repo cache
   - `bodyStruct` infrastructure is already widely used
- Storing current config hash
  - Used to detect config modification by the user. As the user changes the config, the branch won't get rebased (cause of `isBranchModified`) so `isBranchModified` cant be used to detect config changes as it will always be true once other than renovate[bot] have committed to the branch

## Context
- Closes #17152
- closes #11770

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository (GitHub + Bitbucket)

### CODE REVIEW

**Comment by Gabriel-Ladzaretti** [comment]
@rarkins @viceice 
Would appreciate your input on this pull request before I proceed with test coverage 😄

**Comment by Gabriel-Ladzaretti** [comment]
Should we be worried about exceeding 60k characters here?

File: lib/workers/repository/onboarding/branch/rebase.ts [hunk 1]
```diff
@@ -8,26 +9,35 @@ import {
...
+  previousConfigHash?: string
 ): Promise<string | null> {
   logger.debug('Checking if onboarding branch needs rebasing');
+  const configFile = defaultConfigFile(config);
+  const existingContents =
+    (await getFile(configFile, config.onboardingBranch)) ?? '';
+  const currentConfigHash = hasha(existingContents);
+  const contents = await getOnboardingConfigContents(config, configFile);
+
+  if (is.undefined(previousConfigHash)) {
```
Comment (PhilipAbed) [inline lib/workers/repository/onboarding/branch/rebase.ts:27]: why not just (!previousConfigHash) ?

**Review by PhilipAbed — CHANGES_REQUESTED** [review]
MassageMarkdown for other platform that dont support html comments and checkboxes
azure/Bitbucket/bitbucket-server
and soon AWS

File: lib/workers/repository/onboarding/branch/rebase.ts (hunk 1 above)
Comment (Gabriel-Ladzaretti) [reply lib/workers/repository/onboarding/branch/rebase.ts:27]: You are right, we should probably check for null as well.

Style wise I personally prefer `is` as I think it is more readable.

ill use `is.nullOrUndefined` in this case, feel free to object or suggest otherwise per best practices 😀

File: lib/modules/platform/azure/index.ts [hunk 2]
```diff
@@ -747,7 +747,7 @@ export function massageMarkdown(input: string): string {
       'rename PR to start with "rebase!"'
     )
     .replace(regEx(`\n---\n\n.*?<!-- rebase-check -->.*?\n`), '')
-    .replace(regEx(/<!--renovate-debug:.*?-->/), '');
+    .replace(regEx(/<!--renovate-(debug|config-hash):.*?-->/), '');
```
Comment (Gabriel-Ladzaretti) [inline lib/modules/platform/azure/index.ts:750]: Although `debug` and `config-hash` aren't supposed to coexist, but these might be safer.

```suggestion
    .replace(regEx(/<!--renovate-(debug|config-hash):.*?-->/g), '');
```

Or maybe -

```suggestion
    .replace(regEx(/<!--renovate-debug):.*?-->/g), '')
    .replace(regEx(/<!--renovate-config-hash:.*?-->/g), '');
```

File: lib/modules/platform/bitbucket/index.ts [hunk 3]
```diff
@@ -487,7 +487,7 @@ export function massageMarkdown(input: string): string {
     .replace(regEx(/<\/?details>/g), '')
     .replace(regEx(`\n---\n\n.*?<!-- rebase-check -->.*?\n`), '')
     .replace(regEx(/\]\(\.\.\/pull\//g), '](../../pull-requests/')
-    .replace(regEx(/<!--renovate-debug:.*?-->/), '');
+    .replace(regEx(/<!--renovate-(debug|config-hash):.*?-->/), '');
```
Comment (Gabriel-Ladzaretti) [inline lib/modules/platform/bitbucket/index.ts:490]: same

File: lib/modules/platform/pr-body.spec.ts [hunk 4]
```diff
@@ -47,13 +49,50 @@ describe('modules/platform/pr-body', () => {
...
     });
 
-    it('returns rebaseRequested flag', () => {
-      expect(getPrBodyStruct('- [x] <!-- rebase-check -->')).toEqual({
-        hash: '023952693e1e00a52a71b65d9b4804bca6ca9f215c20f6e029dbf420f322d541',
+    it('hashes an undefined body', () => {
+      // nullish operator branch coverage
+      const hash = hashBody(undefined);
+      expect(hash).toBe(hash);
+    });
```
Comment (Gabriel-Ladzaretti) [inline lib/modules/platform/pr-body.spec.ts:56]: As weird as it seems, it covers a previously not cover branch.

File: lib/workers/repository/onboarding/common.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+import { configFileNames } from '../../../config/app-strings';
+import type { RenovateConfig } from '../../../config/types';
+import { logger } from '../../../logger';
+
+export const defaultConfigFile = (config: RenovateConfig): string =>
+  configFileNames.includes(config.onboardingConfigFileName!)
+    ? config.onboardingConfigFileName!
+    : configFileNames[0];
+
+export class OnboardingState {
```
Comment (Gabriel-Ladzaretti) [inline lib/workers/repository/onboarding/common.ts:11]: Although the getter setters here does not hold any logic and a regular variable could suffice, the added benefit of logging on every interaction with it is enough in my view for it to exist.

File: lib/modules/platform/azure/index.ts (hunk 2 above)
Comment (rarkins) [reply lib/modules/platform/azure/index.ts:750]: Always err on the side of keeping it simple / easier maintainability

**Review by rarkins — COMMENTED** [review]
Do you think any of this can be broken up into smaller pieces? There's a lot of code being touched here

File: lib/modules/platform/pr-body.ts [hunk 6]
```diff
@@ -9,6 +10,14 @@ export const prDebugDataRe = regEx(
   /\n?<!--renovate-debug:(?<payload>.*?)-->\n?/
 );
 
+export const renovateConfigHashRe = regEx(
+  /\n?<!--renovate-config-hash:(?<payload>.*?)-->\n?/
+);
+
+export const prCheckboxRe = regEx(
+  /- (?<checkbox>\[[\sx]]) <!-- rebase-check -->/
```
Comment (viceice) [inline lib/modules/platform/pr-body.ts:18]: i don't think we need the extra capture group. what's the benefit? i think a test instead of exec will be faster, as we don't need any position information.

why it's exported?

do we really use the unchecked match?

File: lib/modules/platform/pr-body.ts [hunk 7]
```diff
@@ -28,15 +37,24 @@ export function hashBody(body: string | undefined): string {
...
 }
 
-function isRebaseRequested(body: string | undefined): boolean {
-  return !!body?.includes(`- [x] <!-- rebase-check -->`);
+function isRebaseRequested(body: string): boolean | undefined {
+  const match = prCheckboxRe.exec(body);
+  if (!match) {
+    return undefined;
+  }
+  return match.groups?.checkbox === '[x]';
```
Comment (viceice) [inline lib/modules/platform/pr-body.ts:45]:
```suggestion
  return prCheckboxRe.test(body);
```
this is much easier, see other comment

File: lib/workers/repository/index.ts [hunk 8]
```diff
@@ -40,9 +42,11 @@ export async function renovateRepository(
     logger.debug('Using localDir: ' + localDir);
     config = await initRepo(config);
     addSplit('init');
-    const { branches, branchList, packageFiles } = await extractDependencies(
-      config
-    );
+    const preformExtract =
```
Comment (viceice) [inline lib/workers/repository/index.ts:45]:
```suggestion
    const performExtract =
```
typo?

File: lib/workers/repository/onboarding/branch/index.ts [hunk 9]
```diff
@@ -32,9 +34,23 @@ export async function checkOnboardingBranch(
...
   // global gitAuthor will need to be used
   setGitAuthor(config.gitAuthor);
-  if (await onboardingPrExists(config)) {
+  // TODO #7154
+  const onboardingPr = await getOnboardingPr(config);
+  if (onboardingPr) {
+    const pl = GlobalConfig.get('platform')!;
+    const { rebaseRequested, rawConfigHash } = onboardingPr.bodyStruct ?? {};
+    if (!['github', 'gitlab'].includes(pl)) {
+      logger.debug(`Platform '${pl}' does not support extended markdown`);
```
[the remark this answers is not in the document]
Comment (viceice) [reply lib/workers/repository/onboarding/branch/index.ts:43]: why gitea is unsupported?

File: lib/workers/repository/onboarding/common.ts [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  private static updateRequested = false;
+
+  static get prUpdateRequested(): boolean {
+    logger.debug(
```
Comment (viceice) [inline lib/workers/repository/onboarding/common.ts:15]: i think those should be trace, as you als log something before calling the setter. so reduce the overall log messages as @rarkins already mentioned above

File: lib/workers/repository/onboarding/common.ts [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+  }
+
+  static set prUpdateRequested(value: boolean) {
+    logger.debug({ value }, 'Set OnboardingState.prUpdateRequested');
+    this.updateRequested = value;
+  }
+}
+
+export function toSha256(input: string): string {
+  return hasha(input, { algorithm: 'sha256' });
```
Comment (viceice) [inline lib/workers/repository/onboarding/common.ts:29]: don't explicit define the hash, use default value even it's `sha512`

File: lib/modules/platform/bitbucket/index.ts (hunk 3 above)
Comment (viceice) [inline lib/modules/platform/bitbucket/index.ts:490]:
```suggestion
    .replace(regEx(/<!--renovate-(debug|config-hash):.*?-->/g), '');
```
just in case

File: lib/modules/platform/pr-body.ts [hunk 12]
```diff
@@ -9,6 +10,14 @@ export const prDebugDataRe = regEx(
   /\n?<!--renovate-debug:(?<payload>.*?)-->\n?/
 );
 
+export const renovateConfigHashRe = regEx(
```
Comment (viceice) [inline lib/modules/platform/pr-body.ts:13]: same here, why export?

File: lib/workers/repository/onboarding/pr/index.ts [hunk 13]
```diff
@@ -126,6 +130,13 @@ If you need any further assistance then you can also [request help here](${
   if (is.string(config.prFooter)) {
     prBody = `${prBody}\n---\n\n${template.compile(config.prFooter, config)}\n`;
   }
+
+  const configFile = defaultConfigFile(config);
+  const existingContents =
+    (await getFile(configFile, config.onboardingBranch)) ?? '';
+  const hash = toSha256(existingContents);
+  prBody += `\n<!--renovate-config-hash:${toBase64(hash)}-->`;
```
Comment (viceice) [inline lib/workers/repository/onboarding/pr/index.ts:138]: i don't think we need to encode the Sha to base64, ad it's already a base64 string.
```suggestion
  prBody += `\n<!--renovate-config-hash:${hash}-->`;
```

File: lib/modules/platform/pr-body.ts [hunk 14]
```diff
@@ -45,10 +63,16 @@ export function getPrBodyStruct(
...
   const rebaseRequested = isRebaseRequested(body);
-  if (rebaseRequested) {
+
+  if (!is.undefined(rebaseRequested)) {
     result.rebaseRequested = rebaseRequested;
   }
 
+  const rawConfigHash = getRenovateConfigHashPayload(body);
+  if (rawConfigHash) {
+    result.rawConfigHash = fromBase64(rawConfigHash);
```
Comment (viceice) [inline lib/modules/platform/pr-body.ts:73]:
```suggestion
    result.rawConfigHash = rawConfigHash;
```
see other comment

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/workers/repository/onboarding/common.ts (hunk 11 above)
Comment (rarkins) [reply lib/workers/repository/onboarding/common.ts:29]: But, look at the function name

File: lib/modules/platform/pr-body.ts (hunk 12 above)
Comment (Gabriel-Ladzaretti) [reply lib/modules/platform/pr-body.ts:13]: I was originally intended to use these in tests.
will be removed

File: lib/modules/platform/pr-body.ts (hunk 7 above)
Comment (Gabriel-Ladzaretti) [reply lib/modules/platform/pr-body.ts:45]: Test won't help me to know it there is no checkbox at all.
If we test for `[x]` we will have only binary info about the checked box. i.e. (true for `[x]`, false for `[ ]` and for no checkbox at all)
but I also need this info in order to "migrate" existing onboarding PRs.

This is quite defensive as we can identify this case if there is no embedded config hash comment, but i prefer checking for both.

File: lib/workers/repository/index.ts (hunk 8 above)
Comment (Gabriel-Ladzaretti) [reply lib/workers/repository/index.ts:45]: indeed a typo 😅

File: lib/workers/repository/onboarding/common.ts (hunk 11 above)
Comment (Gabriel-Ladzaretti) [reply lib/workers/repository/onboarding/common.ts:29]: `sha512` would be twise as long, I think we can make do with `sha256` as its shorter and will do for our purpose.

File: lib/workers/repository/onboarding/branch/index.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (Gabriel-Ladzaretti) [reply lib/workers/repository/onboarding/branch/index.ts:43]: missed it, it is supported. not tested though

File: lib/modules/platform/pr-body.ts (hunk 6 above)
Comment (Gabriel-Ladzaretti) [reply lib/modules/platform/pr-body.ts:18]: https://github.com/renovatebot/renovate/pull/17633#discussion_r964494404

File: lib/modules/platform/azure/index.ts (hunk 2 above)
Comment (Gabriel-Ladzaretti) [reply lib/modules/platform/azure/index.ts]: resolved here https://github.com/renovatebot/renovate/pull/17633#discussion_r964373405

**Comment by Gabriel-Ladzaretti** [comment]
Blocked by -
- #17673

**Comment by Gabriel-Ladzaretti** [comment]
An additional feature flag to enable this was added, defaults to false.
If not set the behavior is the same as before.

File: lib/config/options/index.ts [hunk 15]
```diff
@@ -400,6 +400,17 @@ const options: RenovateOptions[] = [
     globalOnly: true,
     mergeable: true,
   },
+  {
+    name: 'onboardingRebaseCheckbox',
+    description:
+      'Set to enabled rebase/retry markdown checkbox for onboarding PRs.',
```
Comment (viceice) [inline lib/config/options/index.ts:406]:
```suggestion
      'Set to enable rebase/retry markdown checkbox for onboarding PRs.',
```

File: lib/workers/repository/index.ts [hunk 16]
```diff
@@ -92,3 +100,12 @@ export async function renovateRepository(
   logger.info({ durationMs: splits.total }, 'Repository finished');
   return repoResult;
 }
+
+// istanbul ignore next: renovateRepository is ignored
```
Comment (viceice) [inline lib/workers/repository/index.ts:104]: do we really need the ignore? we should add tests instead

File: lib/workers/repository/onboarding/branch/index.spec.ts [hunk 17]
```diff
@@ -245,5 +258,75 @@ describe('workers/repository/onboarding/branch/index', () => {
...
+        config.onboardingRebaseCheckbox = true;
+        OnboardingState.prUpdateRequested = false;
+        git.getFileList.mockResolvedValueOnce(['package.json']);
+        platform.findPr.mockResolvedValueOnce(null);
+        rebase.rebaseOnboardingBranch.mockResolvedValueOnce(null);
+      });
+
+      afterAll(() => {
+        delete config.onboardingRebaseCheckbox;
+      });
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/index.spec.ts:274]: why is this needed?

File: lib/workers/repository/onboarding/branch/check.ts [hunk 18]
```diff
@@ -129,6 +129,6 @@ export const isOnboarded = async (config: RenovateConfig): Promise<boolean> => {
   throw new Error(REPOSITORY_CLOSED_ONBOARDING);
 };
 
-export const onboardingPrExists = async (
+export const getOnboardingPr = async (
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/check.ts:132]: should be refactored in separate PR and converted to regular function.

File: lib/workers/repository/index.ts (hunk 16 above)
Comment (Gabriel-Ladzaretti) [reply lib/workers/repository/index.ts:104]: Problem is that the entire repo worker function is being Istanbul ignored which is where this function is used.
I actually tried un-ignoring the repo worker function but it seemed out of scope for this PR as it requires quite a bit of work, maybe in another unrelated PR?

https://github.com/renovatebot/renovate/blob/f4bbf71b0b1d1cc7ea10ea922ed290bcc9ec5b75/lib/workers/repository/index.ts#L22-L23

File: lib/workers/repository/onboarding/branch/index.spec.ts (hunk 17 above)
Comment (Gabriel-Ladzaretti) [reply lib/workers/repository/onboarding/branch/index.spec.ts:274]:
```suggestion
      afterAll(() => {
        config.onboardingRebaseCheckbox = false;
      });
```
maybe?

i wanted to disable it for future tests where this might interfere (being `true`)

File: lib/workers/repository/onboarding/branch/check.ts (hunk 18 above)
Comment (Gabriel-Ladzaretti) [reply lib/workers/repository/onboarding/branch/check.ts:132]: - https://github.com/renovatebot/renovate/pull/17965

File: lib/workers/repository/index.ts (hunk 16 above)
Comment (viceice) [reply lib/workers/repository/index.ts:104]: ok, let's do it in separate PR

File: lib/workers/repository/onboarding/branch/index.spec.ts (hunk 17 above)
Comment (viceice) [reply lib/workers/repository/onboarding/branch/index.spec.ts:274]: checkout the beforeEach, config is always fresh for each test, so this does nothing

File: lib/workers/repository/index.ts (hunk 16 above)
Comment (Gabriel-Ladzaretti) [reply lib/workers/repository/index.ts:104]: - https://github.com/renovatebot/renovate/issues/17993

File: lib/workers/repository/onboarding/branch/rebase.ts [hunk 19]
```diff
@@ -65,3 +73,16 @@ export async function rebaseOnboardingBranch(
...
+function handleOnboardingManualRebase(
+  previousConfigHash: string | undefined,
+  currentConfigHash: string
+): void {
+  if (is.nullOrUndefined(previousConfigHash)) {
+    logger.debug('Missing previousConfigHash bodyStruct prop in onboarding PR');
+    OnboardingState.prUpdateRequested = true;
+  } else if (previousConfigHash !== currentConfigHash) {
+    logger.debug('Onboarding config has been modified by the user');
+    OnboardingState.prUpdateRequested = true;
```
[the remark this answers is not in the document]
Comment (Gabriel-Ladzaretti) [reply lib/workers/repository/onboarding/branch/rebase.ts:86]: the underlying logic haven't changed, so we will only update the PR in this case.

the next block handles that
```ts
  if (config.onboardingRebaseCheckbox) {
    handleOnboardingManualRebase(previousConfigHash, currentConfigHash);
  }

  // TODO #7154
  if (await isBranchModified(config.onboardingBranch!)) {
    logger.debug('Onboarding branch has been edited and cannot be rebased');
    return null;
  }

```

File: lib/workers/repository/onboarding/branch/index.spec.ts [hunk 20]
```diff
@@ -40,6 +42,7 @@ describe('workers/repository/onboarding/branch/index', () => {
       jest.resetAllMocks();
       config = getConfig();
       config.repository = 'some/repo';
+      OnboardingState.prUpdateRequested = false;
```
[the remark this answers is not in the document]
Comment (Gabriel-Ladzaretti) [reply lib/workers/repository/onboarding/branch/index.spec.ts:45]: `false` is indeed the default,
just curious, is this what you mean?

```ts
describe("ensure OnboardingState.prUpdateRequested default value", () => {
    beforeEach(()=>{
      jest.resetModules();
    })

    it('expects default value to be "false"', () => {
      const onboardingState = require('../common').OnboardingState;
      expect(onboardingState.prUpdateRequested).toBeFalse();
    })
  })
```

File: lib/workers/repository/index.ts [hunk 21]
```diff
@@ -53,7 +59,9 @@ export async function renovateRepository(
       const res = await updateRepo(config, branches);
       setMeta({ repository: config.repository });
       addSplit('update');
-      await setBranchCache(branches);
+      if (performExtract) {
+        await setBranchCache(branches); // update branch cache if performed extraction
```
[the remark this answers is not in the document]
Comment (RahulGautamSingh) [reply lib/workers/repository/index.ts:63]: It should. This part of code isn't related to the branch cache changes. 
@Gabriel-Ladzaretti My understanding is that you don't want to call `setBranchCache` when onboarding to prevent the branch info from being stored right?

File: lib/workers/repository/onboarding/branch/index.spec.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (rarkins) [reply lib/workers/repository/onboarding/branch/index.spec.ts:45]: I'm a little lost here. Seems to make sense for default to be false

File: lib/workers/repository/index.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (rarkins) [reply lib/workers/repository/index.ts:64]: Needs confirmation

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/workers/repository/index.ts [hunk 22]
```diff
@@ -40,9 +42,13 @@ export async function renovateRepository(
...
-    const { branches, branchList, packageFiles } = await extractDependencies(
-      config
-    );
+    const performExtract =
+      config.repoIsOnboarded! ||
+      !config.onboardingRebaseCheckbox ||
+      OnboardingState.prUpdateRequested;
+    const { branches, branchList, packageFiles } = performExtract
+      ? await extractDependencies(config)
+      : emptyExtract(config);
```
[the remark this answers is not in the document]
Comment (Gabriel-Ladzaretti) [reply lib/workers/repository/index.ts:53]: optimization, extraction + lookup is skipped when we know we wont update the onboarding pr.

File: lib/workers/repository/onboarding/common.ts [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+import type { RenovateConfig } from '../../../config/types';
+import { logger } from '../../../logger';
+
+export function defaultConfigFile(config: RenovateConfig): string {
+  return configFileNames.includes(config.onboardingConfigFileName!)
+    ? config.onboardingConfigFileName!
+    : configFileNames[0];
+}
+
+export class OnboardingState {
```
Comment (viceice) [inline lib/workers/repository/onboarding/common.ts:12]: I'm thinking about this class. still not sure if this is the best way to handle static state for a repo. 🤔

Comment (Gabriel-Ladzaretti) [reply lib/workers/repository/onboarding/common.ts:12]: other options I can think of are an exported plain constant (but than we lose the tracing, which is the only benefit of the static class) and maybe `memCache`?

**Review by PhilipAbed — DISMISSED** [review]

**Review by rarkins — CHANGES_REQUESTED** [review]
Needs deconflicting

**Comment by Gabriel-Ladzaretti** [comment]
Deconflicted. 
Needs e2e testing, will be marked as ready for review once done.

File: lib/workers/repository/onboarding/common.ts (hunk 23 above)
Comment (viceice) [reply lib/workers/repository/onboarding/common.ts:12]: ok, leave this class but store the value inside memcache, so we ensure reset after each repo.

File: lib/workers/repository/onboarding/common.ts [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+    return this.updateRequested;
+  }
+
+  static set prUpdateRequested(value: boolean) {
+    logger.trace({ value }, 'Set OnboardingState.prUpdateRequested');
+    this.updateRequested = value;
+  }
+}
+
+export function toSha256(input: string): string {
```
Comment (viceice) [inline lib/workers/repository/onboarding/common.ts:29]: should probably moved to utils for reuse?

**Review by viceice — COMMENTED** [review]
need extensive real repo testing

**Review by viceice — DISMISSED** [review]

**Comment by Gabriel-Ladzaretti** [comment]
Ill run some e2e tests on GitHub and Bitbucket again, hopefully in the coming days

**Review by viceice — DISMISSED** [review]
code lgtm 🤷‍♂️

**Comment by Gabriel-Ladzaretti** [comment]
## e2e testing:

### Github  
Running with the `onboardingRebaseCheckbox: false` ->extraction/lookup is preformed on every run, repo cache updated on every run.

Turning `onboardingRebaseCheckbox: true,` with an existing onboarding -> checkbox is added, lookup performed, repo cache updated.

Running again without modifications to the branch -> no lookup, repo cache remains the same.
```
DEBUG: Repository timing splits (milliseconds) (repository=ladzaretti/onboarding)
       "splits": {"init": 9965, "onboarding": 133, "update": 1},

```

Commit to branch -> changes detected -> extraction + lookup performed -> cache updated
```
DEBUG: Onboarding config has been modified by the user (repository=ladzaretti/onboarding)
...
DEBUG: Repository timing splits (milliseconds) (repository=ladzaretti/onboarding)
       "splits": {"init": 9422, "extract": 572, "lookup": 400, "onboarding": 2481, "update": 0},
       "total": 104374
```

Running again -> no extraction or lookup
```
DEBUG: Repository timing splits (milliseconds) (repository=ladzaretti/onboarding)
       "splits": {"init": 9377, "onboarding": 115, "update": 1},
       "total": 33022

```

Manually updating the PR body by deleting the checkbox prompt -> PR is getting fixed
```
DEBUG: No rebase checkbox was found in the onboarding PR (repository=ladzaretti/onboarding)
DEBUG: Repository timing splits (milliseconds) (repository=ladzaretti/onboarding)
       "splits": {"init": 9182, "extract": 600, "lookup": 389, "onboarding": 2084, "update": 1},
       "total": 23035

```

Turning it off by `onboardingRebaseCheckbox: false` => PR restored to its regular form.


### Bitbucket
enabling the feature has no effect.
```
DEBUG: Repository timing splits (milliseconds) (repository=gabriel-ladzaretti/onboarding)
       "splits": {"init": 14282, "extract": 764, "lookup": 496, "onboarding": 187, "update": 1},
       "total": 16094
```
```json
{"name":"renovate","hostname":"Gabriel-Dell5420","pid":18052,"level":10,"logContext":"FAH6g4QH_X-6ucWQ4Jcrh","repository":"gabriel-ladzaretti/onboarding","msg":"Platform 'bitbucket' does not support extended markdown","time":"2023-01-08T21:48:58.411Z","v":0}


```

**Review by viceice — APPROVED** [review]

**Review by rarkins — APPROVED** [review]
